lib/media/track/localmediatrack.js

```diff
@@ -228,24 +244,62 @@ function mixinLocalMediaTrack(AudioOrVideoTrack) {
   };
 }
 
+/**
+ * Restart the given {@link LocalMediaTrack} if it has been inadvertently stopped.
+ * @private
+ * @param {LocalAudioTrack|LocalVideoTrack} localMediaTrack
+ * @returns {function} Clean up listeners attached by the workaround
+ */
 function restartWhenInadvertentlyStopped(localMediaTrack) {
-  let { mediaStreamTrack } = localMediaTrack;
+  const { _log: log, kind } = localMediaTrack;
+  const detectSilence = { audio: detectSilentAudio, video: detectSilentVideo }[kind];
+
+  let { _dummyEl: el, mediaStreamTrack } = localMediaTrack;
   let trackChangeInProgress = null;
 
+  function checkSilence() {
+    // The dummy element is paused, so play it and then detect silence.
+    return el.play().then(() => detectSilence(el)).then(isSilent => {
+      if (isSilent) {
+        log.warn('Silence detected');
+      } else {
+        log.info('Non-silence detected');
+      }
+      return isSilent;
+    }).catch(error => {
+      log.warn('Failed to detect silence:', error);
+    }).finally(() => {
+      // Pause the dummy element again.
+      el.pause();
+    });
+  }
+
   function shouldReacquireTrack() {
     const { _workaroundWebKitBug1208516Cleanup, isStopped, mediaStreamTrack: { muted } } = localMediaTrack;
     const isInadvertentlyStopped = isStopped && !!_workaroundWebKitBug1208516Cleanup;
-    return document.visibilityState === 'visible' && (muted || isInadvertentlyStopped) && !trackChangeInProgress;
+    return Promise.resolve().then(() => {
+      return document.visibilityState === 'visible'
+        && !trackChangeInProgress
+        && (muted || isInadvertentlyStopped || checkSilence());
```

Comment:
`checkSilence` returns promise, but looks like here we are converting it to `boolean`. Shouldn't we wait for the checkSilence Promise?

Reply:
@makarandp0 As per our discussion, I will add a comment explaining the behavior.

Reply:
@manjeshbhargav as you explained offline, I see how it works now. Suggest add some comments or update this to 
```
    return Promise.resolve().then(() => {
      if (document.visibilityState === 'visible'
        && !trackChangeInProgress
        && (muted || isInadvertentlyStopped)) {
          return checkSilence();
        } else {
          return false;
        }
```
to make it obvious